src/transport/PeerMessageCounter.h

```diff
@@ -95,10 +96,26 @@ class PeerMessageCounter
         if (counter <= mSynced.mMaxCounter)
         {
             uint32_t offset = mSynced.mMaxCounter - counter;
+            ;
+
+            if (mIsGroup && (offset > (UINT32_MAX / 2)))
+            {
+                // Different behaviour since Group Counter can roll over
+                // spec 4.5.1.3
+
+                // Possible roll over, set the offset to the smallest of the two
+                uint32_t secondOffset = counter - mSynced.mMaxCounter;
```

Comment:
No, I don't think this is right.  Let's look at a simple case: `mSynced.mMaxCounter == UINT32_MAX`, `counter == 0`.  This should be allowed, per spec: we are just getting the next message, with the counter increased by 1.

In this case we get `offset == UINT32_MAX`, so we enter this block.  We compute `secondOffset`, get 1.  `1 < UINT32_MAX`, so we set `offset` to 1.

But now we use `offset` as an index into our window, so we are looking at whether we ever got a message with counter `UINT32_MAX - 1`.  We probably have, and then we'll reject the new, perfectly valid, message.

I suspect what should happen for counters with rollover instead is that we never do any `<=` compares on the counter values themselves.  Instead, we compute various differences mode 2^32 (which is what we get automatically on uint32_t) and see where those diferences fall.  So maybe start out by computing `counter - mSynced.mMaxCounter`, compare it to the spec's "between 1 and 2^31-1" range (and accept if so).  Then something similar to determine whether we are in the window...

Reply:
Reworked